src/index.js

```diff
@@ -1,135 +1,216 @@
+    this.release = this.getReleasePromise();
   }
 
-  if (typeof originalEntry === 'string') {
-    return [newEntry, originalEntry];
+  /** Returns whether this plugin is in dryRun mode. */
+  isDryRun() {
+    return this.options.dryRun === true;
   }
 
-  if (typeof originalEntry === 'function') {
-    return () =>
-      Promise.resolve(originalEntry()).then(entry => injectEntry(entry, newEntry));
+  /** Creates a new Sentry CLI instance. */
+  getSentryCli() {
+    if (this.isDryRun()) {
+      return {
+        releases: {
+          proposeVersion: () => Promise.resolve('1.0.0-dev'),
+          new: () => Promise.resolve(),
+          uploadSourceMaps: () => Promise.resolve(),
+          finalize: () => Promise.resolve(),
+        },
+      };
+    }
+
+    return new SentryCli(this.options.configFile);
   }
 
-  return newEntry;
-}
+  /**
+   * Returns a Promise that will solve to the configured release.
+   *
+   * If no release is specified, it uses Sentry CLI to propose a version. The
+   * release string is always trimmed.
+   */
+  getReleasePromise() {
+    return (this.options.release
+      ? Promise.resolve(this.options.release)
+      : this.cli.releases.proposeVersion()
+    ).then(version => `${version}`.trim());
+  }
+
+  /** Checks if the given named entry point should be handled. */
+  checkEntry(key) {
+    const { entries } = this.options;
+    if (!entries) {
+      return true;
+    }
+
+    if (typeof entries === 'function') {
+      return entries(key);
+    }
 
-function injectRelease(compiler, versionPromise) {
-  if (typeof compiler.options === 'undefined') {
-    // Gatekeeper because we are not running in webpack env
-    // probably just tests
-    return;
+    if (entries instanceof RegExp) {
+      return entries.test(key);
+    }
+
+    if (Array.isArray(entries)) {
+      return entries.indexOf(key) >= 0;
+    }
+
+    throw new Error(
+      'Invalid `entries` option: Must be an array, RegExp or function'
+    );
   }
-  const changedCompiler = compiler;
-  changedCompiler.options.entry = injectEntry(
-    changedCompiler.options.entry,
-    path.join(__dirname, 'sentry-webpack.module.js')
-  );
-  if (typeof changedCompiler.options.module === 'undefined') {
-    changedCompiler.options.module = {};
+
+  /** Injects the release string into the given entry point. */
+  injectEntry(originalEntry, newEntry) {
+    if (Array.isArray(originalEntry)) {
+      return [newEntry].concat(originalEntry);
+    }
+
+    if (originalEntry !== null && typeof originalEntry === 'object') {
+      const nextEntries = {};
+      Object.keys(originalEntry).forEach(key => {
+        if (this.checkEntry(key)) {
+          nextEntries[key] = this.injectEntry(originalEntry[key], newEntry);
+        }
+      });
+      return nextEntries;
+    }
+
+    if (typeof originalEntry === 'string') {
+      return [newEntry, originalEntry];
+    }
+
+    if (typeof originalEntry === 'function') {
+      return () =>
```

Comment:
We can just return `Promise.resolve`, without wrapping it in an intermediate function

Reply:
Which means that we also have to support Promises in addition to the other kinds of entry records, yes?

Reply:
Ah, this was the reason. Ignore me then.